tests/e2e/linalg_ext_ops/generate_e2e_fa2_tests.py

```diff
@@ -0,0 +1,496 @@
+#!/usr/bin/env python3
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Generator for e2e fa2 tests.
+"""
+
+import argparse
+import enum
+import dataclasses
+import typing
+import math
+
+
+# Data type of kernel entries. The string values must match MLIR data types.
+@enum.unique
+class QueryElemTypeId(enum.Enum):
+    NONE = ""
+    F32 = "f32"
+    F16 = "f16"
+
+
+# Data type of input entries. The string values must match MLIR data types.
+@enum.unique
+class KeyElemTypeId(enum.Enum):
+    NONE = ""
+    F32 = "f32"
+    F16 = "f16"
+
+
+# Data type of input entries. The string values must match MLIR data types.
+@enum.unique
+class ValueElemTypeId(enum.Enum):
+    NONE = ""
+    F32 = "f32"
+    F16 = "f16"
+
+# Data type of input entries. The string values must match MLIR data types.
+@enum.unique
+class ResultElemTypeId(enum.Enum):
+    NONE = ""
+    F32 = "f32"
+    F16 = "f16"
+
+# Enumerates of the collections of shapes that we can generate tests for.
+# The values are the accepted values for the --shapes= flag.
+@enum.unique
+class ShapesId(enum.Enum):
+    SMALL = "small"
+    MEDIUM = "medium"
+    LARGE = "large"
+
+
+# Enumerates ways to construct MLIR tensor types.
+@enum.unique
+class Dynamicity(enum.Enum):
+    STATIC = "static"  # Use fixed values everywhere. Example: tensor<4x6xf32>.
+
+
+# Enumerates ways to initialize input buffer contents.
+@enum.unique
+class InputGenerator(enum.Enum):
+    ZERO = "zero"  # Fill with zeros
+    RANDOM = "random"  # Fill with (deterministic) pseudorandom values.
+
+
+# Enumerates ways to initialize kernel buffer contents.
+@enum.unique
+class KernelGenerator(enum.Enum):
+    ZERO = "zero"  # Fill with zeros
+    RANDOM = "random"  # Fill with (deterministic) pseudorandom values.
+
+@dataclasses.dataclass
+class TestShapeAndScale:
+    batch: int
+    numHeads: int
+    seqLen: int
+    headDim: int
+    scale: float
+
+# Returns the list of TestShape's to use for the collection of shapes
+# identified by shapes_id.
+def get_test_shapes(shapes_id: ShapesId):
+
+    if shapes_id == ShapesId.SMALL:
+        return [
+            TestShapeAndScale(batch=4, numHeads=4, seqLen=1024, headDim=64, scale=1.0),
+        ]
+    if shapes_id == ShapesId.MEDIUM:
+        return [
+            TestShapeAndScale(batch=8, numHeads=4, seqLen=2048, headDim=128, scale=1.0),
+        ]
+    if shapes_id == ShapesId.LARGE:
+        return [
+            TestShapeAndScale(batch=16, numHeads=4, seqLen=4096, headDim=128, scale=1.0),
+            TestShapeAndScale(batch=16, numHeads=4, seqLen=16384, headDim=64, scale=1.0),
+        ]
```

Comment:
I see that this is a draft PR, but I'm curious about how this PR differs from the previous PR with the same name: https://github.com/iree-org/iree/pull/16953

I'm always hesitant to add new testing infrastructure like these generator scripts, since they add layers of indirection and more lines of code to maintain. Can you explain why the existing tests are insufficient and why generated tests for fa2 are worth the costs?

Reply:
The initial plan was to stay with those test cases, but it was switched to the test suite since they were insufficient to test FA2 thoroughly. The test suite covers the four dimensions with big numbers. It also allows you to add more tests with less effort than hard-coded tests.

Reply:
Is there an issue tracking those plans / discussions?

Also follow-up question: is this unique to "flash attention 2", or is it generic to all kinds of "attention"? I'm not clear on where the distinction lies, in either the model input code or the code generation. Wondering if we'll fork this boilerplate code again when the inevitable "flashier attention 3" comes around :P

Reply:
I can open that issue tracker tomorrow and connect it with this PR. The reference implementation is the Attention algorithm, so, to the best of my knowledge, it should work with all the faster and optimized versions.